### DISCUSSION TITLE
connection.start was never received, likely due to a network timeout

### DISCUSSION THREAD

**Discussion by marmiskeje** [rabbitmq/rabbitmq-dotnet-client#1060] [General]
Hi.

We are getting this error in multiple productions many times with RabbitMQ client version 6.0.0:

connection.start was never received, likely due to a network timeout

```
StackTrace:
System.IO.IOException:
   at RabbitMQ.Client.Framing.Impl.Connection.StartAndTune (RabbitMQ.Client, Version=6.0.0.0, Culture=neutral, PublicKeyToken=89e7d7c5feba84ce)
   at RabbitMQ.Client.Framing.Impl.Connection.Open (RabbitMQ.Client, Version=6.0.0.0, Culture=neutral, PublicKeyToken=89e7d7c5feba84ce)
   at RabbitMQ.Client.Framing.Impl.Connection..ctor (RabbitMQ.Client, Version=6.0.0.0, Culture=neutral, PublicKeyToken=89e7d7c5feba84ce)
   at RabbitMQ.Client.Framing.Impl.ProtocolBase.CreateConnection (RabbitMQ.Client, Version=6.0.0.0, Culture=neutral, PublicKeyToken=89e7d7c5feba84ce)
   at RabbitMQ.Client.ConnectionFactory.CreateConnection (RabbitMQ.Client, Version=6.0.0.0, Culture=neutral, PublicKeyToken=89e7d7c5feba84ce)
```

Version 6.1.0/6.2.1 had problems with IPv6 connection that´s why we are using version 6.0.0.
Is this project dead? Is someone using this library in Azure environment? Old library for .NET was great.
What can we do about this? No, this is not a network-related problem. Error is happening too many times, communication is in one virtual network using static IP address for RabbitMQ server.

Thank you for any help.

**Comment by michaelklishin** [comment]
What leads you to believe that "this project is dead"? It's a client that's over a decade old, the protocol is also mature so rate of commit activity will greatly vary from time period to time period.

You claim that "this is not a networking problem". You have to back claims by evidence in this community.
A traffic capture with `tcpdump` or similar and analysis with [Wireshark](https://www.rabbitmq.com/amqp-wireshark.html) will
tell you what is going on.

The error means that after connecting to the target endpoint, the client did not receive a `connection.start` frame from RabbitMQ
within a certain time window. That code path hasn't changed meaningfully in a while.

**Reply by marmiskeje** [reply]
What leads you to believe that "this project is dead"?
- because when we tried to switch to version 6.2.1 or 6.1.0 we experienced 2 issues and we had to use version 6.0.0 and since then there was no new version addressing these issues:
-- First, on some developer computers there was an error connecting to RabbitMQ server because of IPv6 check.
-- Second, there was problem with multiplexing.

You claim that "this is not a networking problem". You have to back claims by evidence in this community.
- We have multiple production environments in multiple Azure subscriptions which are managed by multiple companies. All of them have this issue. Some of them more than others but all of them even when RabbitMQ server is addresed by IP address and in same VNET.

I can´t do tcpdump because services are hosted using PaaS platform and these operations are very limited.

I know that I need more data. I will try to get it.

**Comment by michaelklishin** [comment]
FWIW this is not something that has been actively reported elsewhere, and [according to NuGet](https://www.nuget.org/packages/RabbitMQ.Client/), `6.2.1` is a release with over 6.6M downloads (vs. 1.2M for `6.0.0`). So this very well may be an environment-specific problem.

**Comment by pabermod** [comment]
I'm facing the same issue: Latest rabbitmq with ssl and erlang 
rabbitmq server in a docker network. I'm redirecting connections to my computer 5671 port to the 5671 port of the docker container.
The client is connecting from another computer that is connected to the vpn of my workplace. The client was connecting fine, but  now without changing anything it is generating this exception

One thing to point out is that the internet connetion through the vpn seems to be slow, I had to increase 
tcp_listen_options.send_timeout to avoid getting some errores but still I'm getting this error right now

![image](https://user-images.githubusercontent.com/9059614/190436863-a9a9ff35-6043-4ec6-9dc6-cddf7646dd0d.png)

**Reply by lukebakken** [reply]
> The client was connecting fine, but now without changing anything it is generating this exception

That statement suggests that the issue is external to your client application and RabbitMQ. Something is closing the TCP connection pretty quickly after it is opened - you can see that in the RabbitMQ log file.

* Does running the client application in the same network as RabbitMQ fix the issue? I'm assuming it does.
* Does restarting the VPN connection fix the issue?

**Reply by pabermod** [reply]
It doesn't I'm trting to connect from the localhost ( my computer, where docker is running).  My computer is not conected to the vpn...

I've been lookint at the rabbitmq client class and for some reason in the StartAndTune method,

connectionStart is being set to null when waiting for a value in connectionStartCell.WaitForValue(); almost instantaniously and closing the connection. I'm not closing the connection from the client as the server is logging.

Setting the log in debug is not showing more info

**Reply by lukebakken** [reply]
> It doesn't I'm trting to connect from the localhost ( my computer, where docker is running). My computer is not conected to the vpn...

I don't understand this statement. Are you saying that if you run your application on the same host that your docker container is running, you still see this issue? No VPN involved at all?

**Reply by pabermod** [reply]
Thats right.

**Reply by lukebakken** [reply]
OK so since you can reproduce this issue, provide me what I need to reproduce it! Ideally it would be a GitHub repository I can clone and run `docker compose up` to see the issue. I really can't spend the time guessing what may be the cause.

**Reply by lukebakken** [reply]
@pabermod Note, I'm putting together a docker compose environment you can start with here - https://github.com/lukebakken/rabbitmq-dotnet-client-1060

**Reply by lukebakken** [reply]
@pabermod that repository is a _complete_ environment that will start a .NET producer and consumer sending messages to RabbitMQ over TLS.

Please see my instructions in the README for how to clone, setup and run the project. Let me know if you see connection issues like you report above, and if so, _**what is different in my code vs yours!**_

For what it's worth, I do not see connection errors.

**Reply by pabermod** [reply]
This is something very strange, after doing more testing I still get this error from my local computer to rabbitmq running on my WSL, but other people on the same network can connect using the same client as I do. Maybe something network related in my computer is blocking myself?

**Reply by pabermod** [reply]
Same error is happening when trying to connect from my computer to rabbitmq running on another computer of the same network

**Reply by pabermod** [reply]
When trying to connect to another instance of rabbitmq of another computer I get the following trace in wireshark:

![image](https://user-images.githubusercontent.com/9059614/190618449-fa8079dd-78c4-44a7-a227-c48329d30215.png)

**Reply by pabermod** [reply]
Trying a fresh install of my client fixed the problem. It seems to be related to a dll problem, as I have recompiled my client, so maybe I copied less dlls or incorrect versions. I have multiple binding redirects (one being system.threading.channels) so maybe it was affecting in some way...

Anyway thanks for your time

**Reply by lukebakken** [reply]
Thanks for letting me know, that's a very unique cause.
